File: bindings/fsharp/src/STRling/IR.fs
namespace STRling

open System
open System.Text.Json
open System.Text.Json.Serialization

// Recursive definition of Types and Converters to handle circular dependencies and attributes

type IRClassItemConverter() =
    inherit JsonConverter<IRClassItem>()
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        use doc = JsonDocument.ParseValue(&reader)
        let root = doc.RootElement
        let typeProp = root.GetProperty("ir").GetString()
        match typeProp with
        | "Range" ->
            let f = root.GetProperty("from").GetString()
            let t = root.GetProperty("to").GetString()
            IRClassRange(f, t)
        | "Char" ->
            let c = root.GetProperty("char").GetString()
            IRClassLiteral c
        | "Esc" ->
            let t = root.GetProperty("type").GetString()
            let mutable pElem = Unchecked.defaultof<JsonElement>
            let p = if root.TryGetProperty("property", &pElem) && pElem.ValueKind <> JsonValueKind.Null then Some(pElem.GetString()) else None
            IRClassEscape(t, p)
        | _ -> failwithf "Unknown IRClassItem type: %s" typeProp

    override this.Write(writer: Utf8JsonWriter, value: IRClassItem, options: JsonSerializerOptions) =
        writer.WriteStartObject()
        match value with
        | IRClassRange(f, t) ->
            writer.WriteString("ir", "Range")
            writer.WriteString("from", f)
            writer.WriteString("to", t)
        | IRClassLiteral c ->
            writer.WriteString("ir", "Char")
            writer.WriteString("char", c)
        | IRClassEscape(t, p) ->
            writer.WriteString("ir", "Esc")
            writer.WriteString("type", t)
            match p with Some prop -> writer.WriteString("property", prop) | None -> ()
        writer.WriteEndObject()

and IROpConverter() =
    inherit JsonConverter<IROp>()

    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        use doc = JsonDocument.ParseValue(&reader)
        let root = doc.RootElement
        let typeProp = root.GetProperty("ir").GetString()
        
        match typeProp with
        | "Alt" ->
            let branches = JsonSerializer.Deserialize<IROp list>(root.GetProperty("branches"), options)
            IRAlt branches
        | "Seq" ->
            let parts = JsonSerializer.Deserialize<IROp list>(root.GetProperty("parts"), options)
            IRSeq parts
        | "Lit" ->
            let value = root.GetProperty("value").GetString()
            IRLit value
        | "Dot" -> IRDot
        | "Anchor" ->
            let at = root.GetProperty("at").GetString()
            IRAnchor at
        | "CharClass" ->
            let negated = root.GetProperty("negated").GetBoolean()
            let itemsJson = root.GetProperty("items")
            let items = 
                [ for item in itemsJson.EnumerateArray() do
                    yield JsonSerializer.Deserialize<IRClassItem>(item.GetRawText(), options) ]
            IRCharClass(negated, items)
        | "Quant" ->
            let child = JsonSerializer.Deserialize<IROp>(root.GetProperty("child"), options)
            let min = root.GetProperty("min").GetInt32()
            let maxElem = root.GetProperty("max")
            let max = if maxElem.ValueKind = JsonValueKind.String then maxElem.GetString() else maxElem.GetInt32().ToString()
            let mode = root.GetProperty("mode").GetString()
            IRQuant(child, min, max, mode)
        | "Group" ->
            let capturing = root.GetProperty("capturing").GetBoolean()
            let body = JsonSerializer.Deserialize<IROp>(root.GetProperty("body"), options)
            let mutable nameElem = Unchecked.defaultof<JsonElement>
            let name = if root.TryGetProperty("name", &nameElem) && nameElem.ValueKind <> JsonValueKind.Null then Some(nameElem.GetString()) else None
            let mutable atomicElem = Unchecked.defaultof<JsonElement>
            let atomic = if root.TryGetProperty("atomic", &atomicElem) then atomicElem.GetBoolean() else false
            IRGroup(capturing, body, name, atomic)
        | "Backref" ->
            let mutable idxElem = Unchecked.defaultof<JsonElement>
            let index = if root.TryGetProperty("byIndex", &idxElem) && idxElem.ValueKind <> JsonValueKind.Null then Some(idxElem.GetInt32()) else None
            let mutable nameElem = Unchecked.defaultof<JsonElement>
            let name = if root.TryGetProperty("byName", &nameElem) && nameElem.ValueKind <> JsonValueKind.Null then Some(nameElem.GetString()) else None
            IRBackref(index, name)
        | "Look" ->
            let dir = root.GetProperty("dir").GetString()
            let neg = root.GetProperty("neg").GetBoolean()
            let body = JsonSerializer.Deserialize<IROp>(root.GetProperty("body"), options)
            IRLook(dir, neg, body)
        | _ -> failwithf "Unknown IR type: %s" typeProp

    override this.Write(writer: Utf8JsonWriter, value: IROp, options: JsonSerializerOptions) =
        writer.WriteStartObject()
        match value with
        | IRAlt branches ->
            writer.WriteString("ir", "Alt")
            writer.WritePropertyName("branches")
            JsonSerializer.Serialize(writer, branches, options)
        | IRSeq parts ->
            writer.WriteString("ir", "Seq")
            writer.WritePropertyName("parts")
            JsonSerializer.Serialize(writer, parts, options)
        | IRLit v ->
            writer.WriteString("ir", "Lit")
            writer.WriteString("value", v)
        | IRDot -> writer.WriteString("ir", "Dot")
        | IRAnchor at ->
            writer.WriteString("ir", "Anchor")
            writer.WriteString("at", at)
        | IRCharClass(negated, items) ->
            writer.WriteString("ir", "CharClass")
            writer.WriteBoolean("negated", negated)
            writer.WritePropertyName("items")
            JsonSerializer.Serialize(writer, items, options)
        | IRQuant(child, min, max, mode) ->
            writer.WriteString("ir", "Quant")
            writer.WritePropertyName("child")
            JsonSerializer.Serialize(writer, child, options)
            writer.WriteNumber("min", min)
            if max = "Inf" then writer.WriteString("max", "Inf") else writer.WriteNumber("max", int max)
            writer.WriteString("mode", mode)
        | IRGroup(capturing, body, name, atomic) ->
            writer.WriteString("ir", "Group")
            writer.WriteBoolean("capturing", capturing)
            writer.WritePropertyName("body")
            JsonSerializer.Serialize(writer, body, options)
            match name with Some n -> writer.WriteString("name", n) | None -> ()
            if atomic then writer.WriteBoolean("atomic", true)
        | IRBackref(index, name) ->
            writer.WriteString("ir", "Backref")
            match index with Some i -> writer.WriteNumber("byIndex", i) | None -> ()
            match name with Some n -> writer.WriteString("byName", n) | None -> ()
        | IRLook(dir, neg, body) ->
            writer.WriteString("ir", "Look")
            writer.WriteString("dir", dir)
            writer.WriteBoolean("neg", neg)
            writer.WritePropertyName("body")
            JsonSerializer.Serialize(writer, body, options)
        writer.WriteEndObject()

and [<JsonConverter(typeof<IRClassItemConverter>)>] IRClassItem =
    | IRClassRange of fromCh: string * toCh: string
    | IRClassLiteral of ch: string
    | IRClassEscape of type_: string * property: string option

and [<JsonConverter(typeof<IROpConverter>)>] IROp =
    | IRAlt of branches: IROp list
    | IRSeq of parts: IROp list
    | IRLit of value: string
    | IRDot
    | IRAnchor of at: string
    | IRCharClass of negated: bool * items: IRClassItem list
    | IRQuant of child: IROp * min: int * max: string * mode: string
    | IRGroup of capturing: bool * body: IROp * name: string option * atomic: bool
    | IRBackref of byIndex: int option * byName: string option
    | IRLook of dir: string * neg: bool * body: IROp

File: bindings/ruby/lib/strling/ir.rb
# frozen_string_literal: true

require_relative 'core/nodes'

module Strling
  module IR
    Lit = Data.define(:ir, :value)
    Quant = Data.define(:ir, :child, :min, :max, :mode)
    CharClass = Data.define(:ir, :negated, :items)
    Char = Data.define(:ir, :char)
    Group = Data.define(:ir, :capturing, :body, :name, :atomic)
    Alt = Data.define(:ir, :branches)
    Esc = Data.define(:ir, :type, :property)
    Seq = Data.define(:ir, :parts)
    Look = Data.define(:ir, :dir, :neg, :body)
    Anchor = Data.define(:ir, :at)
    BackRef = Data.define(:ir, :byIndex, :byName)
    Range = Data.define(:ir, :from, :to)
    Dot = Data.define(:ir)
    # Add others as needed

    class Compiler
      def self.compile(ast_node)
        return nil if ast_node.nil?

        case ast_node
        # Core Nodes
        when Strling::Core::Lit
          Lit.new(ir: 'Lit', value: ast_node.value)
        when Strling::Core::Quant
          Quant.new(
            ir: 'Quant',
            child: compile(ast_node.child),
            min: ast_node.min,
            max: ast_node.max,
            mode: ast_node.mode
          )
        when Strling::Core::Group
          Group.new(
            ir: 'Group',
            capturing: ast_node.capturing,
            body: compile(ast_node.body),
            name: ast_node.name,
            atomic: ast_node.atomic
          )
        when Strling::Core::Alt
          Alt.new(
            ir: 'Alt',
            branches: ast_node.branches.map { |b| compile(b) }
          )
        when Strling::Core::Seq
          parts = ast_node.parts.map { |p| compile(p) }
          merged_parts = []
          parts.each do |part|
            if part.is_a?(Lit) && merged_parts.last.is_a?(Lit)
              last = merged_parts.pop
              merged_parts << Lit.new(ir: 'Lit', value: last.value + part.value)
            else
              merged_parts << part
            end
          end
          if merged_parts.size == 1
            merged_parts.first
          else
            Seq.new(ir: 'Seq', parts: merged_parts)
          end
        when Strling::Core::Esc
          Esc.new(ir: 'Esc', type: ast_node.type, property: ast_node.property)
        when Strling::Core::Anchor
          at = ast_node.at == 'NonWordBoundary' ? 'NotWordBoundary' : ast_node.at
          Anchor.new(ir: 'Anchor', at: at)
        when Strling::Core::Look
          Look.new(
            ir: 'Look',
            dir: ast_node.dir,
            neg: ast_node.neg,
            body: compile(ast_node.body)
          )
        # Legacy Nodes
        when Strling::Nodes::Literal
          Lit.new(ir: 'Lit', value: ast_node.value)
        when Strling::Nodes::Quantifier
          mode = if ast_node.possessive
                   'Possessive'
                 elsif ast_node.lazy
                   'Lazy'
                 else
                   'Greedy'
                 end
          
          # Handle max being null/nil -> "Inf"
          max_val = ast_node.max.nil? ? 'Inf' : ast_node.max

          Quant.new(
            ir: 'Quant',
            child: compile(ast_node.target),
            min: ast_node.min,
            max: max_val,
            mode: mode
          )
        when Strling::Nodes::CharacterClass
          items = ast_node.members.map do |member|
            if member.is_a?(Strling::Nodes::Literal)
              Char.new(ir: 'Char', char: member.value)
            elsif member.is_a?(Strling::Nodes::UnicodeProperty)
               type = member.negated ? 'P' : 'p'
               Esc.new(ir: 'Esc', type: type, property: member.value)
            elsif member.is_a?(Strling::Nodes::Escape)
               compile(member)
            else
              # Fallback or recursive compile if members can be other things
              compile(member)
            end
          end
          CharClass.new(ir: 'CharClass', negated: ast_node.negated, items: items)
        when Strling::Nodes::Group
          Group.new(
            ir: 'Group',
            capturing: ast_node.capturing,
            body: compile(ast_node.body),
            name: ast_node.name,
            atomic: ast_node.atomic ? true : nil
          )
        when Strling::Nodes::Alternation
          Alt.new(
            ir: 'Alt',
            branches: ast_node.alternatives.map { |a| compile(a) }
          )
        when Strling::Nodes::UnicodeProperty
          # If it appears standalone (not in CharClass), it might still be Esc?
          # In unicode_property_short.json it was inside CharacterClass.
          # Let's assume it maps to Esc.
          type = ast_node.negated ? 'P' : 'p'
          Esc.new(ir: 'Esc', type: type, property: ast_node.value)
        when Strling::Nodes::Sequence
          parts = ast_node.parts.map { |p| compile(p) }
          
          # Merge adjacent Lit nodes
          merged_parts = []
          parts.each do |part|
            if part.is_a?(Lit) && merged_parts.last.is_a?(Lit)
              last = merged_parts.pop
              merged_parts << Lit.new(ir: 'Lit', value: last.value + part.value)
            else
              merged_parts << part
            end
          end

          if merged_parts.size == 1
            merged_parts.first
          else
            Seq.new(ir: 'Seq', parts: merged_parts)
          end
        when Strling::Nodes::Lookahead
          Look.new(
            ir: 'Look',
            dir: 'Ahead',
            neg: ast_node.negative || false, # AST might not have negative if it's a separate node type "NegativeLookahead"
            body: compile(ast_node.body)
          )
        when Strling::Nodes::Lookbehind
          Look.new(
            ir: 'Look',
            dir: 'Behind',
            neg: ast_node.negative || false,
            body: compile(ast_node.body)
          )
        when Strling::Nodes::Anchor
          at = ast_node.at == 'NonWordBoundary' ? 'NotWordBoundary' : ast_node.at
          Anchor.new(ir: 'Anchor', at: at)
        when Strling::Nodes::Escape
          type = case ast_node.kind
                 when 'digit' then 'd'
                 when 'not-digit' then 'D'
                 when 'space' then 's'
                 when 'not-space' then 'S'
                 when 'word' then 'w'
                 when 'not-word' then 'W'
                 # Add others if needed
                 else ast_node.kind # Fallback
                 end
          Esc.new(ir: 'Esc', type: type, property: nil)
        when Strling::Nodes::BackReference
          BackRef.new(
            ir: 'Backref',
            byIndex: ast_node.index,
            byName: ast_node.name
          )
        when Strling::Nodes::Range
          Range.new(ir: 'Range', from: ast_node.from, to: ast_node.to)
        when Strling::Nodes::Dot
          Dot.new(ir: 'Dot')
        else
          # Fallback
          # puts "Warning: Unknown AST node for compilation: #{ast_node.class}"
          nil
        end
      end
    end
  end
end
